ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_similarity/TextSimilarityTranslator.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2023 Aryn
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.opensearch.ml.engine.algorithms.text_similarity;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.opensearch.ml.common.output.model.MLResultDataType;
+import org.opensearch.ml.common.output.model.ModelTensor;
+import org.opensearch.ml.common.output.model.ModelTensors;
+import org.opensearch.ml.engine.algorithms.SentenceTransformerTranslator;
+
+import ai.djl.huggingface.tokenizers.Encoding;
+import ai.djl.modality.Input;
+import ai.djl.modality.Output;
+import ai.djl.ndarray.NDArray;
+import ai.djl.ndarray.NDList;
+import ai.djl.ndarray.NDManager;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.translate.TranslatorContext;
+
+public class TextSimilarityTranslator extends SentenceTransformerTranslator {
+
+    @Override
+    public NDList processInput(TranslatorContext ctx, Input input) {
+        String sentence = input.getAsString(0);
+        String context = input.getAsString(1);
+        NDManager manager = ctx.getNDManager();
+        NDList ndList = new NDList();
+        Encoding encodings = tokenizer.encode(sentence, context);
+        long[] indices = encodings.getIds();
+        long[] attentionMask = encodings.getAttentionMask();
+        long[] tokenTypes = encodings.getTypeIds();
+
+        NDArray indicesArray = manager.create(indices);
+        indicesArray.setName("input_ids");
+
+        NDArray attentionMaskArray = manager.create(attentionMask);
+        attentionMaskArray.setName("attention_mask");
+
+        NDArray tokenTypeArray = manager.create(tokenTypes);
+        tokenTypeArray.setName("token_type_ids");
+
+        ndList.add(indicesArray);
+        ndList.add(attentionMaskArray);
+        ndList.add(tokenTypeArray);
+        return ndList;
+    }
+
+    @Override
+    public Output processOutput(TranslatorContext ctx, NDList list) {
+        Output output = new Output(200, "OK");
+
+        List<ModelTensor> outputs = new ArrayList<>();
+        Iterator<NDArray> iterator = list.iterator();
+        while (iterator.hasNext()) {
+            NDArray ndArray = iterator.next();
+            String name = ndArray.getName();
```

Comment:
In your output (issue description), it's showing `logits`.  This doesn't seem quite meaningful. May be we can add function name here?

Reply:
In addition, does this only support torchscript or both (torchscript & onnx)?

Reply:
haven't tried with ONNX - never compiled an ONNX model before, so I don't know. But I _think_ I didn't do anything that would preclude it?

Reply:
ack to logits. changing to similarity in next push, but I wanna resolve the function name thing first

Reply:
If you look at the text embedding algorithm, we needed to add pooling layer information for onnx translator. So can we try to check if this will also work with onnx format or not?

Reply:
k. looks like onnx doesn't work. (can upload the model but the output is messed up). will enforce TORCH_SCRIPT format here and support ONNX in the future, sg?

Reply:
the end of the response:
```
      {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [
            128
          ],
          "data": [
            0.9998161,
            -0.9851868,
            ...
            0.99590653,
            -0.18241836
          ],
          "byte_buffer": {
            "array": "8/N/PzQ1fL+s5n8/pvl0v1c7cD+h7HM/MpJ+P0KlST+1HHu/iD5vvy3Tfj/JEnS/CiV3vwAAgL9ulXM/Bf16P7vuRj+1AXu/SF9RPyUwfD/dVn+/5XZ0P73ybL8xt3+/vOx/vygae7/pgH4/Aad/P++yTD+IfnC/dQ9pv9yTfT9zwXm/TP9/vzL/fz85ln+/rGhMv6D9cL8U3n2//Tp4P5bjYj9GYXK/HSN/P8mkf78AAIA/KVJ+P+jtfj/C4X+/kxh/v0oTeT8SOnC/RE13v3NFfj9uJHQ/DQZ5PwAAgL8AAIA/A1x7v4yXfz8lqn8/fxt9v5byfr9LdX0/1ntzPwiqeT/3/38//bV/P68HYz/X73a/hal+P6O3f78cYXw/LFZ/Pzm1fL92+32/mIF8v+mwRb+OXHo/kPt7v31Ofb/PeXw/p1t3PxeBfL+Z/nU/tex/v3r8fj/hn3+/8kl8P0e9dT/ms3E/AACAP1t2db9ckni/DUZ4P//afb+AlXw/Jt1/P8qUf79v1n8/drd7v7FnV7+FXl+/8j9/P6uYfj/X3Hm/tWB6P5SvbT9Uwn8/Hudlv39Ffr9hmHw/TNp1P7nyfz8rtH8/AACAP8Anfj/RQXk/4S99P1bme79591e/kwN+v/r3fz83r36/Wmx+P9LIfz8i5G8/u/N+P+HLOr4=",
            "order": "LITTLE_ENDIAN"
          }
        }
      ]
    }
  ]
}
```
(my terminal cut off the top).
But yes, we'll probably need to add some custom onnx output translator to select the right entry in this tensor list (there were like nine of them? But varying sizes so maybe it gives you the hidden states is that possible?) and possibly convert to single number somehow. I would have thought the classification head would get pulled along with the model at compile time so hopefully it won't be too bad. I guess it's also possible my compile script was bad...
```
def _prepare_cross_encoder_onnx(model_name):
    tk = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForSequenceClassification.from_pretrained(model_name)
    mname = Path(model_name).name
    
    tk.save_pretrained(f"/tmp/{mname}-tokenizer")
    transformers.convert_graph_to_onnx.convert(
        framework="pt",
        model=model_name,
        output=Path(f"/tmp/onnx/{mname}.onnx"),
        opset=15,
    )
    r = requests.get("https://github.com/opensearch-project/opensearch-py-ml/raw/main/LICENSE")

    with ZipFile(f"/tmp/{mname}.zip", "w") as f:
        f.write(f"/tmp/onnx/{mname}.onnx", arcname=f"{mname}.onnx")
        f.write(f"/tmp/{mname}-tokenizer/tokenizer.json", arcname="tokenizer.json")
        f.writestr("LICENSE", r.content)
    shutil.rmtree(f"/tmp/{mname}-tokenizer")
    shutil.rmtree("/tmp/onnx/")
    return f"/tmp/{mname}.zip"
```
btw are we allowed to just slap an Apache license on all these models? That seems fishy to me

Reply:
aha I guess it does drop the classification head, since it's just running off of the name. lemme see if there's a way to onnx-ify with that piece. (might be able to simplify the embedding models too?)

Reply:
ok, if we use `torch.onnx.export` then onnx works just fine. I think the issue is that `transformers.convert_graph_to_onnx.convert` drops the classification head at the end of the model. I think this would also fix embedding models so we don't have to do the pooling by hand.

Reply:
Cool, thanks for deep diving here. Could you please cut an issue [here](https://github.com/opensearch-project/opensearch-py-ml/issues) so that we can change this method to use `torch.onnx.export`. We should change that method as I remember `convert_graph_to_onnx` is going to be deprecated too.

Reply:
https://github.com/opensearch-project/opensearch-py-ml/issues/347